### PR TITLE
Scroll messages in all lock dialogs into view

### PR DESCRIPTION
**Pull request by m-idler**
Messages might appear outside of users viewport. This PR fixes this by scrolling global messages into the users viewport. Hiding the users keyboard doesn't fix the problem because of the huge amount of different resolutions and display ratios.
see https://github.com/auth0/lock/issues/959

### CODE REVIEW

**Comment by luisrudge** [comment]
Can you make a gif comparing both results?

**Comment by m-idler** [comment]
Current state - you won't recognise any errors that occur
state-current
changed state - lock dialog will scroll to the position of the error message
state-new

In the previous commit we added a parameter ```scrollGlobalMessagesIntoView``` to not brake/change current behaviour. By default  ```scrollGlobalMessagesIntoView``` is disabled.

File: src/core/index.js [hunk 1]
```diff
@@ -183,7 +186,8 @@ export const ui = {
   popupOptions: lock => getUIAttribute(lock, 'popupOptions'),
   primaryColor: lock => getUIAttribute(lock, 'primaryColor'),
   authButtonsTheme: lock => getUIAttribute(lock, 'authButtonsTheme'),
-  rememberLastLogin: m => tget(m, 'rememberLastLogin', getUIAttribute(m, 'rememberLastLogin'))
+  rememberLastLogin: m => tget(m, 'rememberLastLogin', getUIAttribute(m, 'rememberLastLogin')),
+  scrollGlobalMessagesIntoView: lock => getUIAttribute(lock, 'scrollGlobalMessagesIntoView')
```
Comment (luisrudge) [inline src/core/index.js]: shouldn't this be `tget(m, 'scrollGlobalMessagesIntoView', getUIAttribute(m, 'scrollGlobalMessagesIntoView')),`?

File: src/core/index.js [hunk 2]
```diff
@@ -211,7 +215,8 @@ function extractAuthOptions(options) {
     sso,
     state,
     nonce
-  } = options.auth || {};
+  } =
```
Comment (luisrudge) [inline src/core/index.js]: no need to break line here

**Comment by luisrudge** [comment]
That's amazing. Thanks for the PR 🎉

File: src/core/index.js (hunk 1 above)
Comment (fstoerkle) [inline src/core/index.js]: what's the difference? actually, I am not really sure what `tget`,`tgetUI`, etc are doing. However, all properties above don't use `tget`, so I would assume the current way to retrieve the option is sufficient.

Comment (luisrudge) [inline src/core/index.js]: Yup. you're right. I was just looking at the rememberLastLogin option, but it is a special case.

**Comment by luisrudge** [comment]
Alright, we're good. I just requested some feedback from our ui/ux designer. I'll wait for him before merging this. Thaaaaaaanks for the PR 👍

File: src/core/index.js (hunk 1 above)
Comment (fstoerkle) [inline src/core/index.js]: 👍

File: src/core/index.js (hunk 2 above)
Comment (m-idler) [inline src/core/index.js]: Seems like this was changed by prettier

**Comment by m-idler** [comment]
Add to small additional commits to fix one of the new tests and add some documentation for the new parameter to the README

**Comment by luisrudge** [comment]
Hey, can you rebase this? THanks!

**Comment by beneliflo** [comment]
I just tested and what I saw is the scroll it varies depends of the height. This happen too with you?

lock_scroll

**Comment by luisrudge** [comment]
I think you have to refresh after resizing the height, no?

**Comment by beneliflo** [comment]
@luisrudge with refresh. Same:

lock_scroll_2

**Comment by luisrudge** [comment]
thanks @beneliflo. Can you take a look @m-idler?

**Comment by m-idler** [comment]
I can reproduce the mentioned behaviour with small viewport heights too. After playing around with it a little bit, it seems that it's not caused by the scrolling itself but by the focus being set to the first input later.
In chrome.jsx I changed the timeout value from 17 to 1000 for testing:
```javascript
if (!prevProps.error && error) {
      const input = this.findAutofocusInput();
      if (input) {
        // TODO clear timeout
        setTimeout(() => input.focus(), 1000);
      }
      return;
}
```
For chrome, it seems that the focused input is vertically-center-scrolled into the viewport, in firefox it's just scrolled at any position into the viewport. You can see the behaviour of Chrome and Firefox in the attached gifs.

Im not sure if it's a (good) solution, to skip setting the focus when ```scrollGlobalMessagesIntoView``` is enabled

chrome-focus
firefox-focus

**Comment by luisrudge** [comment]
Thanks @m-idler. Just to give you a feedback, I'm discussing internally with our design team if we should keep the scroll-to-input behaviour at all. I'm also discussing if we should make `scrollGlobalMessagesIntoView` default to `true` instead of `false`. I'll push to your branch once I receive the final answer.

**Comment by luisrudge** [comment]
Thanks @m-idler for all your effort. I made a small change to make this the default behaviour and removing the scroll-to-input behaviour-on-error that you pointed out. Starting from the next release, `scrollGlobalMessagesIntoView` will be true by default. 🎉

**Comment by luisrudge** [comment]
🎉 :shipit: 🚀

**Pull request merged**